lib/membrane_ffmpeg_swscale/scaler.ex

```diff
@@ -0,0 +1,97 @@
+defmodule Membrane.FFmpeg.SWScale.Scaler do
+  @moduledoc """
+  This element performs video scaling, using SWScale module of FFmpeg library.
+
+  There are two options that have to be passed:
+  - `output_width` - desired scaled video width.
+  - `output_height` - desired scaled video height.
+
+  Scaling consists of two operations:
+  - scaling itself - resizing video frame with keeping original ratio. After that operation at least one of the dimensions of the input frame match the respective dimension of the desired output size. The second one (if does not match) is smaller than its respective dimension.
+  - adding paddings - if one dimension does not match after scaling, paddings have to be added. They are put on both sides of the scaled frame equally. They are either above and below the frame or on the left and right sides of it. It depends on the dimension that did not match after scaling.
+
+  The element needs input in the YUV420p format. It processes one frame at a time. The element also requires
+  getting caps with video width and height, so either `Membrane.Element.RawVideo.Parser` or `Membrane.H264.FFmpeg.Decoder`
+  is recommended to be present before Scaler in the pipeline.
+
+  The output of the element is also in the YUV420p format. It has the size as specified in the options. All caps different than
+  width and height are passed unchanged to the next element in the pipeline.
+  """
+  use Membrane.Filter
+  alias __MODULE__.Native
+  alias Membrane.Buffer
+  alias Membrane.Caps.Video.Raw
+
+  def_options output_width: [
+                type: :int,
+                description: "Width of the scaled video."
+              ],
+              output_height: [
+                type: :int,
+                description: "Height of the scaled video."
+              ]
+
+  def_input_pad :input,
+    demand_unit: :buffers,
+    caps: {Raw, format: :I420, aligned: true}
+
+  def_output_pad :output,
+    caps: {Raw, format: :I420, aligned: true}
+
+  @impl true
+  def handle_init(options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.put(:native_state, nil)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :buffers, _context, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: payload} = buffer, _context, state) do
+    %{native_state: native_state} = state
+
+    case native_state do
+      nil ->
+        raise(RuntimeError, :uninitialized_native_state)
+
+      _state ->
+        with {:ok, frame} <- Native.scale(payload, native_state) do
+          buffer = [buffer: {:output, %{buffer | payload: frame}}]
+
+          {{:ok, buffer}, state}
+        else
+          {:error, reason} -> {{:error, reason}, state}
+        end
+    end
+  end
```

Comment:
Linter would complain about too much nesting. Instead of case here, do the pattern matching in a function clause